File: server/internal/messages/receiver.go
package messages

import (
	"bufio"
	"context"
	"encoding/json"
	"log/slog"
	"net"
	"os"
	"rinha/internal/payments"
	"rinha/internal/payments/workers"
	"sync"
	"time"
)

type Receiver struct {
	socketPath string
	ln         net.Listener
	logger     *slog.Logger
	workers    *workers.WorkerPool

	ctx    context.Context
	cancel context.CancelFunc
	wg     sync.WaitGroup
}

func NewReceiver(socketPath string, workers *workers.WorkerPool, logger *slog.Logger) *Receiver {
	ctx, cancel := context.WithCancel(context.Background())
	return &Receiver{
		socketPath: socketPath,
		ctx:        ctx,
		cancel:     cancel,
		workers:    workers,
		logger:     logger,
	}
}

func (r *Receiver) Start() error {
	_ = os.Remove(r.socketPath)

	ln, err := net.Listen("unix", r.socketPath)

	if err != nil {
		return err
	}
	r.ln = ln

	r.wg.Add(1)
	r.acceptLoop()
	return nil
}

func (r *Receiver) Stop() {
	r.cancel()
	_ = r.ln.Close()
	r.wg.Wait()
}

func (r *Receiver) acceptLoop() {
	defer r.wg.Done()

	for {
		conn, err := r.ln.Accept()
		if err != nil {
			r.logger.Error("Failed to accept connection", "err", err)
			select {
			case <-r.ctx.Done():
				return
			default:
				time.Sleep(10 * time.Millisecond)
				continue
			}
		}

		r.logger.Info("Accepted UDS connection", "remoteAddr", conn.RemoteAddr())
		r.wg.Add(1)
		go r.readProducer(conn)
	}
}

func (r *Receiver) readProducer(conn net.Conn) {
	defer r.wg.Done()
	defer conn.Close()

	sc := bufio.NewScanner(conn)
	for sc.Scan() {
		var msg payments.PaymentMessage
		if err := json.Unmarshal(sc.Bytes(), &msg); err != nil {
			continue
		}

		select {
		case <-r.ctx.Done():
			return
		default:
			r.logger.Info("Received message", "msg", msg)
			r.workers.Submit(&msg)
		}
	}
}

File: worker/src/receiver.rs
use crate::worker_pool::WorkerPool;
use std::sync::Arc;
use std::time::Duration;
use bytes::Bytes;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::net::{UnixListener, UnixStream};
use tokio::sync::Semaphore;

pub struct Receiver {
    socket_path: String,
    workers: Arc<WorkerPool>,
    conn_sem: Arc<Semaphore>
}

#[derive(Debug)]
pub enum ReceiverError {
    SocketError(std::io::Error),
}

impl std::fmt::Display for ReceiverError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ReceiverError::SocketError(e) => write!(f, "Socket error: {}", e),
        }
    }
}

impl std::error::Error for ReceiverError {}

impl Receiver {
    pub fn new(socket_path: String, workers: Arc<WorkerPool>) -> Self {
        Self {
            socket_path,
            workers,
            conn_sem: Arc::new(Semaphore::new(512))
        }
    }

    pub async fn start(&mut self) -> Result<(), ReceiverError> {
        tracing::info!("Starting receiver");
        if std::fs::metadata(&self.socket_path).is_ok() {
            let _ = std::fs::remove_file(&self.socket_path);
        }

        let listener = UnixListener::bind(&self.socket_path).map_err(ReceiverError::SocketError)?;

        if let Err(e) = std::fs::set_permissions(
            &self.socket_path,
            std::os::unix::fs::PermissionsExt::from_mode(0o600),
        ) {
            tracing::warn!(error = %e, "Failed to set permissions on socket");
        }

        self.accept_loop(listener).await;

        Ok(())
    }

    async fn accept_loop(&self, listener: UnixListener) {
        let workers = Arc::clone(&self.workers);

        tracing::info!("Listening on {}", self.socket_path);

        loop {
            tracing::debug!("Waiting for connection");
            match listener.accept().await {
                Ok((stream, addr)) => {
                    tracing::info!(?addr, "Accepted UDS connection");

                    let workers_clone = Arc::clone(&workers);
                    let semaphore = Arc::clone(&self.conn_sem);

                    tokio::task::spawn(async move {
                        let _permit = semaphore.acquire().await.unwrap();
                        Self::read_producer(stream, workers_clone).await;
                    });
                }
                Err(e) => {
                    tracing::error!(error = %e, "Failed to accept connection");
                    tokio::time::sleep(Duration::from_millis(10)).await;
                }
            }
        }
    }

    async fn read_producer(stream: UnixStream, workers: Arc<WorkerPool>) {
        let mut reader = BufReader::with_capacity(8192, stream);
        let mut buffer = Vec::with_capacity(1024);

        loop {
            match reader.read_until(b'\n', &mut buffer).await {
                Ok(0) => {
                    tracing::info!("Read producer disconnected");
                    return;
                }
                Ok(_) => {
                    if buffer.last() == Some(&b'\n') {
                        buffer.pop();
                    }

                    if !buffer.is_empty() {
                        let bytes = Bytes::copy_from_slice(&buffer);
                        if let Err(e) = workers.submit(bytes).await {
                            tracing::warn!(error = %e, "Failed to submit message to worker pool");
                        }
                    }

                    buffer.clear();
                }
                Err(e) => {
                    tracing::warn!(error = %e, "Error reading from connection");
                    return;
                }
            }
        }
    }
}
